### PR TITLE
Dependencies

### PR DESCRIPTION
**Pull request by moorek**

### CODE REVIEW

**Comment by tjhei** [comment]
Looks good so far. Next step is to remove viscosity_depends_on?

**Comment by moorek** [comment]
Yeah that would be next.

File: include/aspect/material_model/interface.h [hunk 1]
```diff
@@ -693,6 +681,24 @@ namespace aspect
...
+         * output variables (such as viscosity, density, etc) depend
+         * on the input variables pressure, temperature, strain rate,
+         * and compositional fields.
+         *
+         * The constructor of this class calls the default
+         * constructor of this member variable which in turn
+         * initializes the object to invalid values. Derived classes
+         * then need to fill it either in their constructor (if they
+         * already know the correct dependences at that time) or
+         * at the end of their parse_parameter() functions where
```
Comment (gassmoeller) [inline include/aspect/material_model/interface.h]: Does it need to be the parse_parameters() function? I would think, this is a case for the initialize() function.

Comment (bangerth) [inline include/aspect/material_model/interface.h]: Either is fine, I suppose. We know by the time we leave parse_parameters what our parameters are. Not all material models have an initialize function, but all have a parse parameters function.

File: source/material_model/interface.cc [hunk 2]
```diff
@@ -163,6 +175,16 @@ namespace aspect
     }
 
 
+
```
Comment (gassmoeller) [inline source/material_model/interface.cc]: Remove this empty line please, the other functions have 2 empty lines between them.

**Comment by gassmoeller** [comment]
Great work :+1: I like this simplification a lot.

Please introduce the old information about the dependencies into the new structure, e.g. CompositionReaction::density_depends_on., DiffusionDislocation::density_depends_on / thermal_conductivity_depends_on, LatentHeat::density_depends_on, same for LatentHeatMelt, MorencyDoin, MultiComponent::density_depends_on, Simple::density_depends_on, SimpleCompressible::density_depends_on, Steinberger:: all functions, Table

File: include/aspect/material_model/interface.h (hunk 1 above)
Comment (tjhei) [inline include/aspect/material_model/interface.h]: It makes sense that we do it in the same place everywhere even though it is technically possible to do it in a different place.

Comment (bangerth) [inline include/aspect/material_model/interface.h]: It's not of concern to me either way. @moorek -- feel free to change the text here if you do this in the `initialize` functions.

**Comment by moorek** [comment]
wait to merge until Wolfgang's test is merged.

File: source/material_model/composition_reaction.cc [hunk 3]
```diff
@@ -373,6 +310,43 @@ namespace aspect
...
       prm.leave_subsection();
+
+//dependencies
+//viscosity
+// depends on temperature
+      if ( (thermal_viscosity_exponent != 0))
+        this->model_dependence.viscosity = NonlinearDependence::temperature | NonlinearDependence::compositional_fields;
+// doesn't depend on temperature
+      else
+        this->model_dependence.viscosity = NonlinearDependence::compositional_fields;
```
Comment (tjhei) [inline source/material_model/composition_reaction.cc]: it only depends on compositional_fields if ``` (composition_viscosity_prefactor_1 != 1.0 || composition_viscosity_prefactor_2 != 1.0)```

File: source/material_model/composition_reaction.cc [hunk 4]
```diff
@@ -373,6 +310,43 @@ namespace aspect
...
+
+//density dependencies
+//depends on temperature
+      if (         (thermal_alpha != 0))
```
Comment (tjhei) [inline source/material_model/composition_reaction.cc]: you can write the whole if-then-else block as:
```
this->model_dependence.density = NonlinearDependence::none;
if (thermal_alpha != 0)
  this->model_dependence.density |= NonlinearDependence::temperature;
if (compositional_delta_rho_1 != 0 || compositional_delta_rho_2 != 0)
  this->model_dependence.density |= NonlinearDependence::compositionn;
 
```

File: tests/material_model_dependencies.cc [hunk 5]
```diff
@@ -22,72 +22,72 @@ namespace aspect
 	 const aspect::MaterialModel::Interface<dim> &model
 	   = this->get_material_model();
 	 std::cout << "viscosity depends on "
-		   << (model.viscosity_depends_on (pressure)
+		   << ((model.get_model_dependence().viscosity | MaterialModel::NonlinearDependence::pressure)
```
Comment (tjhei) [inline tests/material_model_dependencies.cc]: The operator needs to be ```&``` not ```|``` here and below.

File: tests/shear-thinning.cc [hunk 6]
```diff
@@ -43,12 +43,12 @@ namespace aspect
...
-    bool
-    ShearThinning<dim>::
-    viscosity_depends_on (const NonlinearDependence::Dependence dependence) const
+template <int dim>
+    void
+   ShearThinning<dim>::parse_parameters (ParameterHandler &prm)
     {
-      return ((dependence & NonlinearDependence::strain_rate) != NonlinearDependence::none);
+      Simple<dim>::parse_parameters(prm);
+      this->model_dependence.viscosity = MaterialModel::NonlinearDependence::strain_rate;
```
Comment (tjhei) [inline tests/shear-thinning.cc]: this needs to be ```|=```

**Comment by tjhei** [comment]
Please see comments above. Looks very good otherwise.

**Comment by moorek** [comment]
WIP, just changing things to see if it passes the tests. do not merge

**Comment by tjhei** [comment]
Any updates on this? Is the version here your most recent work? If you can't find the time to finish it, somebody could continue where you left off. Thanks!

**Comment by moorek** [comment]
Hi Timo, I tried to edit it at home but can't figure out why all of the tests are failing since the dependencies should be correctly written out (possibly with one or two exceptions modified to try to see why the tests were failing). I don't think I'll be able to continue this, if somebody else wants to take a look.

**Comment by gassmoeller** [comment]
Hi @moorek, do not worry, I am happy to take a look. After all this is a community project, not an exam, so let us know, if you get stuck somewhere (in particular also your own projects, not only this "clean up somebody else's mess" project).

In particular after a first check I think it was just a tiny bug. The simple material model produced a wrong dependency (it claimed to be viscosity-dependent on composition all the time). Because of that dependency the stokes preconditioner got rebuild for all models. 
The results are still the same (no harm done by rebuilding the preconditioner, except for slowing down some models), but now they create an additional line of output ("Rebuild Stokes preconditioner ..."), which caused the tests to fail.
I will create a new pull request (still containing your commits), to force a new run of the tester.

**Comment by gassmoeller** [comment]
Since #565 is merged, this one can be closed.

**Pull request closed without merging**